File: app/src/main/java/com/disruption/cookcentral/network/RecipeApiServiceProvider.java
package com.disruption.cookcentral.network;

import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

import static com.disruption.cookcentral.utils.Constants.BASE_URL;

public class RecipeApiServiceProvider {
    private static Retrofit retrofit;
    private static HttpLoggingInterceptor sHttpLoggingInterceptor =
            new HttpLoggingInterceptor().setLevel(HttpLoggingInterceptor.Level.BASIC);

    private static OkHttpClient sOkHttpClient = new OkHttpClient.Builder()
            .addInterceptor(sHttpLoggingInterceptor)
            .build();

    private static Retrofit retrofitInstance() {
        if (retrofit == null) {
            retrofit = new Retrofit.Builder()
                    .baseUrl(BASE_URL)
                    .addConverterFactory(GsonConverterFactory.create())
                    .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                    .client(sOkHttpClient)
                    .build();
        }
        return retrofit;
    }

    public static RecipeApiService getRecipeApiService() {
        return retrofitInstance().create(RecipeApiService.class);
    }
}

File: app/src/main/java/com/disruption/cookcentral/network/RecipeApiServiceProvider.kt
package com.disruption.cookcentral.network

import com.disruption.cookcentral.utils.Constants
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory

object RecipeApiServiceProvider {

    private var retrofit: Retrofit? = null

    private fun retrofitInstance(): Retrofit? {
        if (retrofit == null) {
            retrofit = Retrofit.Builder()
                    .baseUrl(Constants.BASE_URL)
                    .addConverterFactory(GsonConverterFactory.create())
                    .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                    .build()
        }
        return retrofit
    }

    @JvmStatic
    val recipeApiService: RecipeApiService
        get() = retrofitInstance()!!.create(RecipeApiService::class.java)
}